### ISSUE TITLE
Linux modding tutorial (Wine, Proton, Lutris)...maybe?

### ISSUE THREAD

**Issue by junglized** [tsunamods-codes/7th-Heaven#19]
It's not really a bug but rather tutorial request. I'm on Pop!_OS using Lutris for gaming. Everything works just fine. Unfortunately I can't figure out how to run 7th Heaven with any version of FF7. Could anyone provide a simple tutorial on how to mod FF VII with Wine or Steam + Ptoton version?

Thanks a lot!

**Comment by julianxhokaxhiu** [comment]
Hi @junglized,

Making 7th working on Linux is currently a painful process. If you can manage to make it work under Wine, then you should be good to go on other layers using it. Porting it natively wouldn't make much sense as you'd need Wine to play the game anyway and hooking the Linux memory process would be too cumberstone.

Consider trying steps described in here to see if you can make it work: https://appdb.winehq.org/objectManager.php?sClass=version&iId=38203

Let us know if you succeed.

Cheers,
Julian

**Comment by junglized** [comment]
I wish 7th Heaven was tested more thoroughly with Wine (or Steam Proton) on Linux to make it usable for those, who choose this solution over proprietary systems, such as Windows.

Best regsrds!

**Comment by sjenkins7** [comment]
I stumbled across this, as I was interested in getting modded FF7 running under Linux, as it's my daily driver.

Unfortunately, currently there is no way to do so, due to the disc check on launch, due to missing implementation in Mono: https://github.com/mono/mono/blob/bdd772531d379b4e78593587d15113c37edd4a64/mcs/class/corlib/System.IO/DriveInfo.cs#L109

In order to stand a chance in getting it working, you would need to:

1. Get this `VolumeLabel` implemented in Mono for Windows
2. Push for this to be released in both Mono & wine-mono

For the record, I was able to get as far as this point using the following ( Ubuntu with the Steam FF7 installed ):

```bash
#!/bin/bash

export WINEPREFIX=/home/${USER}/.local/share/wineprefixes/ff7-7th-heaven
FF7DIR="/home/${USER}/.steam/debian-installation/steamapps/common/FINAL FANTASY VII/" # Assuming a default Steam install here - feel free to change as needed
FHDIR="${WINEPREFIX}/drive_c/Program Files/7th Heaven"

# Add PPA for cdemu
sudo add-apt-repository ppa:cdemu/ppa
sudo apt-get update
sudo apt install cdemu-client -y # Swap for gcdemu if you would rather also have a GUI

# Setup prefix
wine64 wineboot
winetricks dxvk # unsure if needed yet - usually a safe bet for the game itself
winetricks corefonts
winetricks winxp # work around file menus crashing using the Vista+ version

# Install wine-mono - .NET 4.8 in wine is horribly broken
wget https://github.com/madewokherd/wine-mono/releases/download/wine-mono-7.1.1/wine-mono-7.1.1-x86.msi
wine64 msiexec /i wine-mono-7.1.1-x86.msi
rm wine-mono-7.1.1-x86.msi

# Run installer
wget https://files.tsunamods.com/7ThHeavenSetup.exe
wine64 7ThHeavenSetup.exe # The installer will crash at the end, due to being unable to create a shortcut - the core install still works

# make a few dirs that don't exist but show up in the guide
mkdir -p "${FF7DIR}/mods/Textures"
mkdir -p "${FF7DIR}/mods/7th Heaven"

# Load the Disc into cdemu
cdemu load 0 "${FHDIR}/Resources/FF7DISC1.ISO"
# Mount the disc itself, so WINE can auto-detect it
pause

# Run it
wine64 "${WINEPREFIX}/drive_c/Program Files/7th Heaven/7th Heaven.exe"

```

This would get you as far as launching 7th Heaven and doing basic configuration, but without being able to pass the CD Check, there is very little anyone can do.

For the Tsunamods Team: If we were able to get this resolved elsewhere first, would you be open to offering this as an option? I'd be more than willing to spend the time to provide support, to make it happen.

**Comment by barkermn01** [comment]
>```
># Load the Disc into cdemu
>cdemu load 0 "${FHDIR}/Resources/FF7DISC1.ISO"
># Mount the disc itself, so WINE can auto-detect it
>pause
>```

Okies, so for Linux, I will add a setting that is "DoNotMount" AKA 7th heaven won't attempt to automount the disk. that should get around that problem for you, we should have that as an option anyway so if someone wants they can call a Partition or something else that name and so don't have to have it mounted

**Comment by junglized** [comment]
After unsuccessful attempts of running 7th Heaven with Wine or Proton I ditched it and moved to The Reunion installed with Protontricks. Works out of the box. No so many mods but the most relevant for me - SYW and Ninostyle characters - are present for and function without a problem.

**Comment by barkermn01** [comment]
> unsuccessful attempts of running 7th Heaven with Wine or Proton I ditched it and moved to The Reunion

Would you be willing to test the new 7th heaven when it's released it should detect reunion and handle it and launch the version without reunion, and then when the game exits undo the changes so that reunion works again?

**Comment by junglized** [comment]
> > unsuccessful attempts of running 7th Heaven with Wine or Proton I ditched it and moved to The Reunion
> 
> Would you be willing to test the new 7th heaven when it's released it should detect reunion and handle it and launch the version without reunion, and then when the game exits undo the changes so that reunion works again?

So, in other words, would you like me to try installing it over The Reunion with Steam + Proton version of FF VII (Linux Mint 20.3 + Steam/Proton + FF VII + The Reunion is my current environment/game setup)?

**Comment by sjenkins7** [comment]
After getting stuck in this evening & trying to get further - looks like there is more to be done under Wine.

I was able to patch my own install locally to return the correct disc label, which got the game launched as a vanilla install.

Once you try to add mods, things fall apart around EasyHook. First it needs  a`System.Runtime.Remoting.Channels.Ipc.IpcServerChannel` constructor that is missing from the Mono distribution.

If you get past that - FF7 simply crashes. Unsure at this point if its the hooks in 7th Heaven, EasyHook itself or missing stuff in wine / mono. 
```
23:24:30|INFO|SeventhHeaven.Classes.GameLauncher|Copying EasyHook.dll to FF7 path (if not found or older version detected) ...
23:24:30|INFO|SeventhHeaven.Classes.GameLauncher|       skipped copying EasyHook.dll ...
23:24:30|INFO|SeventhHeaven.Classes.GameLauncher|       skipped copying EasyHook32.dll ...
23:24:30|INFO|SeventhHeaven.Classes.GameLauncher|Checking if Reunion mod is installed ...
23:24:30|INFO|SeventhHeaven.Classes.GameLauncher|       found: False
23:24:30|INFO|SeventhHeaven.Classes.GameLauncher|Launching additional programs to run (if any) ...
23:24:31|INFO|SeventhHeaven.Classes.GameLauncher|Writing temporary runtime profile file to C:\users\simon\Temp\tmp7e19183.tmp ...
23:24:31|INFO|SeventhHeaven.Classes.GameLauncher|Attempting to inject with EasyHook ...
23:25:24|WARN|SeventhHeaven.Classes.GameLauncher|       received errors: STATUS_INTERNAL_ERROR: Unknown error in injected assembler code. (Code: 5);  ...
```
![image](https://user-images.githubusercontent.com/20030065/153092714-fb8c0946-d08f-438c-a7b6-d6ee6de11d50.png)

**Comment by sjenkins7** [comment]
> Okies, so for Linux, I will add a setting that is "DoNotMount" AKA 7th heaven won't attempt to automount the disk. that should get around that problem for you, we should have that as an option anyway so if someone wants they can call a Partition or something else that name and so don't have to have it mounted

While appreciated, not sure if it's necessary beyond just failing fast.

WinCDEmu doesn't seem to work at all in Wine, but have not really had time to look into why

The main problem is that the stub simply returns the path. In Wine, this means it would be the drive letter itself - nothing could ever match without fixing / hacking Mono.

**Comment by sjenkins7** [comment]
After a bit more experimentation with different launchers, was able to get the base game to load under 7th Heaven without mods.

Using [Bottles](https://usebottles.com):

* Create New Gaming Bottle
* Install Steam under "Installers"
* Install `dotnet48` under "Dependencies"
* Enable "use bottles runtime" under "Preferences"
* Downgrade DXVK to 1.8.1 under "Preferences" ( to avoid graphical glitches that will make install difficult )
* Install FF7 via Steam
* Run 7H Installer via "Run Executable"
* Mount CD as before ( if you are using the Flatpak version of Bottles, you need to make sure it has permission to access the disk )

At this point, it's using the real .NET 4.8 & a proper Steam install location, so everything runs correctly as before. 

So far I've only been able to get the game to reliably launch with DX11, so make sure this is selected in the launcher.

I don't get any SFX, only the BGM - entirely possible I'm missing some config here, but haven't had time to dig into that in more detail

Sadly running with mods doesn't appear to be possible still - as soon as `EasyHook` gets involved, the game crashes (albeit with a different stack trace this time):
![image](https://user-images.githubusercontent.com/20030065/156264836-7171e4c0-e254-4c4f-813a-bbee511b15ff.png)

it appears that SHA isn't implemented in WINE 7.2? ( the version used at the time of this comment)

Not being familiar with the hooking system being used - is it possible to avoid this?

Overall, the game was far more stable - Was able to play through as far as escaping the first reactor, without any real issues, etc.

**Comment by barkermn01** [comment]
> it appears that SHA isn't implemented in WINE 7.2? ( the version used at the time of this comment)

strange because it's listed in the source as being there, https://source.winehq.org/WineAPI/A_SHAFinal.html

**Comment by sjenkins7** [comment]
Finally got a break through - the fork of WINE being used was at fault. Switched to a different runner & got it working immediately. Tested a few minutes in with Cosmo Memory & Cosmo FMV without issue.

The config I used:
```
Arch: win64
Creation_Date: '2022-03-05 12:15:48.176125'
Custom_Path: false
DLL_Overrides: {}
DXVK: dxvk-1.9.4
Environment: Gaming
Environment_Variables: {}
Installed_Dependencies:
- dotnet40
- dotnet48
- d3dx9
- msls31
- riched20
- allfonts
- d3dcompiler_43
- d3dcompiler_47
- vcredist2019
Latest_Executables: []
NVAPI: dxvk-nvapi-v0.5.2
Name: 7th Heaven
Parameters:
    aco_compiler: false
    custom_dpi: 96
    discrete_gpu: true
    dxvk: true
    dxvk_hud: false
    dxvk_nvapi: false
    fixme_logs: false
    fsr: false
    fsr_level: 5
    fullscreen_capture: false
    gamemode: false
    gamescope: false
    gamescope_borderless: false
    gamescope_fps: 0
    gamescope_fps_no_focus: 0
    gamescope_fullscreen: true
    gamescope_game_height: 0
    gamescope_game_width: 0
    gamescope_scaling: false
    gamescope_window_height: 0
    gamescope_window_width: 0
    pulseaudio_latency: true
    sync: esync
    take_focus: false
    use_runtime: true
    virtual_desktop: false
    virtual_desktop_res: 1280x720
    vkd3d: false
Path: 7th-Heaven
Runner: vaniglia-6.23
State: 9
Uninstallers:
    allfonts: false
    d3dcompiler_43: false
    d3dcompiler_47: false
    d3dx9: false
    dotnet40: Microsoft .NET Framework 4 Extended
    dotnet48: true
    msls31: false
    riched20: false
    vcredist2019: true
Update_Date: '2022-03-05 13:18:20.710904'
VKD3D: vkd3d-proton-2.5
Versioning: true
Windows: win10
WorkingDir: ''
```

Save this config & import it in Bottles. Should then just be a case of:
* Install Steam
* Install FF7
* Disable DXVK ( for the duration of setting up 7th Heaven )
* Install 7th Heaven
* Configure Mods & set DX11
* Enable DXVK ( only when you intend to play )

and enjoy:
![image](https://user-images.githubusercontent.com/20030065/156886771-7da34bcd-f429-4dae-9c6e-90cd10934768.png)

**Comment by julianxhokaxhiu** [comment]
Great progress over there!

What do you think about creating a Bottle Installer recipe to make it easy for other users? https://maintainers.usebottles.com/installers/Introduction

Would be really nice to have 7th Heaven in the list. Thank you in advance!

**Comment by sjenkins7** [comment]
> What do you think about creating a Bottle Installer recipe to make it easy for other users? https://maintainers.usebottles.com/installers/Introduction

Can certainly look into doing one. On the face of it, they don't appear to need to be silent installs, which should make it easier. That said - still need to get Steam to install FF7 in the first place.

**Comment by sjenkins7** [comment]
On the face of it - creating the installer might not be practical yet.

Given the Steam version appears to be the golden standard for installs & there doesn't appear to be any way to chain installers, it looks like we would have to maintain our own Installer for Steam as part of this.

With Bottles releasing experimental support for Steam Proton fixes - would be interesting to reach out to the devs, to see if there are plans to support this. In our case, all we really need is FF7 + the last bits needed to get 7th heaven.

**Comment by julianxhokaxhiu** [comment]
If you're up for it, feel free to approach them and follow up this topic. The easier we manage to get this up and running on Linux, the better. Keep us updated and thank you in advance for the time spent on this!

**Comment by SeongGino** [comment]
Apologies, but I have to interject here a little; I seem to be running into a similar (albeit not identical) issue @sjenkins7 was having with EasyHook.

In my case (Arch, Kernel 5.17, KDE 5.42), it's a standalone prefix created specifically for 7H with Lutris (only needs `dotnet48 vcrun2019`) so I've gotten to the point of the mod manager running and detecting/configuring mods, and running vanilla FF7 works as well (Game Driver/FFNx Stable set to use Vulkan, and the prefix has a virtual D:\ set to a random directory and given the label *FF7DISC1*).
![2022_04-04 01-37-54](https://user-images.githubusercontent.com/7321839/161482757-85e5a3cb-80dd-4b62-b2fe-ab0f639f330d.png)

Unfortunately, running *with* mods fails at EasyHook with 'Code 5'.
![2022_04-04 01-37-13](https://user-images.githubusercontent.com/7321839/161482731-341ee225-be2e-4c9c-8dfe-992563a95889.png)
But there is no visible Wine crash/debugger window appearing, and running the instance from a terminal gives the most generic and least helpful line prints I've ever seen.
![2022_04-04 02-24-15](https://user-images.githubusercontent.com/7321839/161485337-7e286d6f-5f75-4b45-9e0e-e91aba26e51f.png)

This has been tried with System (mainline) Wine 7.5, lutris' 7.2 fshack, 6.21 fshack, and GE 7-7; all of which exhibit the same symptoms. I've tried using the same Bottles-sourced fork used in the referenced comments above, but the prefix won't even launch with it using Lutris - and I've too many other Windows/GOG games setup with it to even consider switching frontends.

___

On a related note, the Bottles idea is nice, but it does leave out what's probably a small but existing minority that uses the original 1998 4-CD release. fwiw, it's clearly much _less_ of a hassle to get FF7 running without WineSteam on top of it.

**Comment by sjenkins7** [comment]
> Unfortunately, running with mods fails at EasyHook with 'Code 5'. 

This was in line with what I saw when putting the notes together. The only way i managed to get it working, was with the bottle's distributed `vaniglia` variant. Anything else tripped with this same error & a crash of the FF7 process. It's odd that you aren't getting the debugger show up though

If someone was able to narrow this down, we could figure out where to ask for support, be it a patch in WINE or a fix to a bad hook. At very least we have access to the source of both 7th Heaven & EasyHook, which can help. I can go rummaging around the 7th heaven source and try stuff out, but I'm not sure how much success there would be, as I'm not a WINE developer

> On a related note, the Bottles idea is nice, but it does leave out what's probably a small but existing minority that uses the original 1998 4-CD release

While I was obviously putting these notes together under the guise of the Steam version, I believe the CD version could be substituted - you wouldn't need to mount the ISO needed for the Steam version, because you already have the original disc. Unfortunately I don't have one to hand to prove either way.

**Comment by julianxhokaxhiu** [comment]
You could try enabling the `Code 5 Fix` flag in `Settings -> Game Launcher` and see how it goes.

The code 5 fix is required even on Windows depending on various circumstances. Unfortunately it's a legacy thing we have from EasyHook and we can't change that, not in the near future.

**Comment by SeongGino** [comment]
> You could try enabling the `Code 5 Fix` flag in `Settings -> Game Launcher` and see how it goes.
> 
> The code 5 fix is required even on Windows depending on various circumstances. Unfortunately it's a legacy thing we have from EasyHook and we can't change that, not in the near future.

This I have tried, but it initially caused 7H to crash outright (Object Not Set to a Reference-type deal). Unfortunately, looking at 7H's source, all the Code 5 fix does is apply a regkey for HiDPI compatibility(?) - a setting that does not apply to Wine at all (it just acts like it's 100% scale). Adding the same reg key path in the prefix gets past the crash, but (ironically) causes the same Code 5 error regardless.

And @sjenkins7, my rambling was not intended to frame you in an undesirable spotlight and I hope it didn't come off as such. It is just strange that Bottles' fork specifically is content with EasyHook, but not any other popular upstream/downstream Wine variant - and that it has to seemingly throw a tantrum when used elsewhere outside of its own program driiives me up a small wall.

**Comment by sjenkins7** [comment]
> And @sjenkins7, my rambling was not intended to frame you in an undesirable spotlight and I hope it didn't come off as such. It is just strange that Bottles' fork specifically is content with EasyHook, but not any other popular upstream/downstream Wine variant - and that it has to seemingly throw a tantrum when used elsewhere outside of its own program driiives me up a small wall.

Don't worry - wasn't taken as that :)

For what it's worth, I used to use Lutris a lot before i rebuild my setup on Endeavour recently ( after putting the above notes together ). Always intended to swing back around for this - once I have something 100% working all of the time, I can isolate the differences in the prefixes. This might help get things working across the board.

Also worth noting the installer logic for Bottles made it into a release. I will take a look at trying to automate as much of the install as I can.

**Comment by barkermn01** [comment]
So from a technical standpoint normally if you are getting read error's at 0x0 addresses, the application using the memory addresses does not have application offset applied, Normally Applications in an OS is provided with an offset at the start and all memory addresses are Offset+internal position, Easyhook needs to know the memory offset location for the ff7 exe, and normally windows hold this against the application process (in the process manager) so 7th heaven provides the process ID to EasyHook, EasyHook can then talk to the Windows Process Manager to get it's memory offset if your getting 0x0 addresses across the board it means this is failing and getting a NULL response as NULL = 0x0

**Comment by SeongGino** [comment]
@sjenkins7 Well, good to know we both share a common distro together, so we can rule out another possible variable.

@barkermn01 I tried your VS build - even going as far as making a new 32-bit prefix specifically for it with the same changes I noted before - but unfortunately it made no difference. Same symptoms and output as before.

**Comment by sjenkins7** [comment]
An update - I have the start of an installer for bottles available at: https://github.com/sjenkins7/bottles-programs

I'd encourage everyone interested to test this out and let me know if you have any issues. Please use a `custom` environment, rather than the `Gaming` environment, to select the correct runtime

A few things to be aware of:

If you are using Steam with Steam Guard - you might find you need to tweak settings to get it working ( https://github.com/bottlesdevs/programs/issues/57 )

FF7 is not automatically installed - Either the 1998 CDs or via Steam

Recommended settings to tweak once installed ( should we look to make these the default if running in wine? ):

* Game Driver ( will need to launch vanilla for the first time before you can set these ):
* * set `Graphics API` to `DirectX 12`( Game appears to always crash without this )
* * Set `Music Option` to `VGMStream` ( No music without this )
* Game Launcher:
* * `Do not auto mount Disk` (This is non-functional)
* * Set `Auto-Mount Game Disk` to `off`
* * Set `Auto-Dismount Game Disk` to `off`


Full details can be found at: https://github.com/sjenkins7/bottles-programs/blob/feature/7th-heaven/Reviews/7th-heaven.md

**Comment by SeongGino** [comment]
> set Graphics API to DirectX 12( Game appears to always crash without this )

@sjenkins7 Shouldn't it be able to run fine with the `Vulkan` API? It worked for me when running FF7 on its own with Nx.

**Comment by sjenkins7** [comment]
> @sjenkins7 Shouldn't it be able to run fine with the Vulkan API? It worked for me when running FF7 on its own with Nx.

@SeongGino  I found that it was frequently causing lockups, usually before you get into the 1st reactor. This often locks up X11 as well for a few seconds, though it recovers ( the game itself doesn't and has to be killed ). At first i thought it was a texture mod causing it, but I was still hitting this without any mods.

So far DX12 hasn't caused any issues for me - tested as far as the second reactor.

**Comment by SeongGino** [comment]
> > @sjenkins7 Shouldn't it be able to run fine with the Vulkan API? It worked for me when running FF7 on its own with Nx.
> 
> @SeongGino I found that it was frequently causing lockups, usually before you get into the 1st reactor. This often locks up X11 as well for a few seconds, though it recovers ( the game itself doesn't and has to be killed ). At first i thought it was a texture mod causing it, but I was still hitting this without any mods.
> 
> So far DX12 hasn't caused any issues for me - tested as far as the second reactor.

That sounds more like a driver crash. Especially since DX12 over VKD3D has an extra overhead that limits GPU utilization...

**Comment by sjenkins7** [comment]
> That sounds more like a driver crash. Especially since DX12 over VKD3D has an extra overhead that limits GPU utilization...

Most likely. In my case, i'm favouring the stability - Even when rocking a fair amount of mods in my current test run, its sticking solidly to the 15/30/60 FPS depending on the area. No sudden drops etc, even with an older card ( i run this on a GTX 1070 ).

![image](https://user-images.githubusercontent.com/20030065/163728675-646a37d5-e214-4cca-a38a-22a24c549bce.png)

Worth noting that I didn't enable VKD3D in the installer either - so far it's proven unnecessary to do so

**Comment by sjenkins7** [comment]
Small update for everyone:

I've pushed a new version of the installer - hopefully people can try it out and let me know how it works.

Mounting the CD isn't needed any more - the installer scripts the setup of an empty FF7 CD to satisfy the game. Messing with the mount settings is also not needed.

Install guide for this version:

* Create `Custom` Bottle
* Install `Steam`
* Login to steam and install FF7 ( you may bump into https://github.com/bottlesdevs/Bottles/issues/1401 )
* Install `7th Heaven`
* Launch game for the first time ( crash )
* set Graphics API to DirectX 12

I opened https://github.com/tsunamods-codes/7th-Heaven/pull/85 to streamline the install. The installer will automatically launch 7th heaven without this, which isn't desired ( because we try to install FF7 directly before this, it will launch with no installed FF7 & you have to fill in the directory config yourself otherwise ). Once the 2 remaining Bottles bugs get resolved, I can streamline this into an (almost) 1-click install ( I can't auto login & grab FF7 from Steam sadly! )

The game still hard crashes on first launch, because you still need to specify DX12. Not sure what we can do to specify this in advance? Can't fill this in because FFNx.cfg doesn't exist on first launch?

**Comment by julianxhokaxhiu** [comment]
Great progress so far, thank you for pushing hard on this topic. Ideally once this process has been streamlined we could theoretically find a way to add 7th on the SteamDeck library. Looking forward into this.

> The game still hard crashes on first launch, because you still need to specify DX12. Not sure what we can do to specify this in advance?

You could automate the installation of the latest FFNx stable using the [manual installation instructions](https://github.com/julianxhokaxhiu/FFNx/blob/master/docs/how_to_install.md#1998-eidos-release), change the [backend renderer parameter](https://github.com/julianxhokaxhiu/FFNx/blob/master/misc/FFNx.toml#L11-L20) to what you desire and 7th will automatically handle the rest on launch.

I also merged the PR, thank you again.

**Comment by sjenkins7** [comment]
> You could automate the installation of the latest FFNx stable using the [manual installation instructions](https://github.com/julianxhokaxhiu/FFNx/blob/master/docs/how_to_install.md#1998-eidos-release), change the [backend renderer parameter](https://github.com/julianxhokaxhiu/FFNx/blob/master/misc/FFNx.toml#L11-L20) to what you desire and 7th will automatically handle the rest on launch.

Not sure how practical this would be - I would be replicating a lot of 7th Heavens logic into a system that doesn't have this kind of capability yet:

* How can I guarantee the install directory exists? Steam doesn't provide this, outside of [steamcmd](https://developer.valvesoftware.com/wiki/SteamCMD#Windows), which would only be viable if I could download FF7 without an account
* Installers are done via exe / msi - there is no current capability to unpack a ZIP
* * This could be done via `dependencies`, but then the first point applies - dependencies are installed before any installer steps are executed.

Other things I have to consider ( though admittedly, I could do this as a separate installer ):
* What if the user wants to use the CDs?
* Where is the install directory? Particularly important if i need to cater for multiple releases

While I can update config as part of the install steps, either:
*  `FFNx.toml` would need to exist after running the 7th Heaven installer ( maybe 7th heaven needs to download and manage FFNx as part of install, not when launching? )
* I would have to bundle a pre-made `FFNx.toml`, which may cause more problems if it starts deviating.

**Comment by lumnie** [comment]
Hello. Been spending hours attempting to get this process down to work with the Steam Deck and been having trouble trying to get this customer installer working in bottles. I've attempted the cd command listed above to copy below but it seems I'm still unable to get the customer installer to load up in the installer's section. Prior to this, I tried the earlier suggestion of getting 7th Heaven to work by using the script to make a custom bottle but was running into issues with the disc mount feature causing me to start over.

Can someone guide me as to how this CD command works to import 7th Heaven to bottles as an installer? Originally, I've just been using the + to add the installer that way.

**Comment by sjenkins7** [comment]
> Can someone guide me as to how this CD command works to import 7th Heaven to bottles as an installer? Originally, I've just been using the + to add the installer that way.

Because I haven't submitted the installer yet to the bottles maintainers, those wanting to test it in advance have to jump through additional hoops. The commands I mentioned before tell bottles to use my repository for installers, instead of the official one.

Assuming the Steam Deck doesn't change the install locations or have tools missing, the instructions provided should be enough. Unfortunately I don't have one to check, so I'd have to rely on screenshots etc..

Once the installer has been put to review & then accepted by the bottles maintainers, it will show up as any other installer without any of these extra steps. This is purely if you wish to test the installer beforehand

**Comment by lumnie** [comment]
Seem to be running into the issue with the 7th heaven mounting the disco or even mounting the virtual disco when i swapped options as per mentioned earlier in the thread. The Bottle recipe calls for install steam then FF7 then 7th heaven. Given that all usually takes place within a fake C drive, does that mean that I would need to have FF7 be placed in the C drive created in bottles?

Given steam deck has it's built in steam application, it's made with it's own fake drive which can be accessed with bottles. Do you think that may be the reason the application is having trouble mounting a fake ISO? Or is there a separate setting that's suppose to help with that in the preferences? 

![ksnip_20220527-193154](https://user-images.githubusercontent.com/101153429/170800232-dbbcf58b-41a9-465c-b667-346b63f2abd5.png)

**Comment by julianxhokaxhiu** [comment]
> Given steam deck has it's built in steam application, it's made with it's own fake drive which can be accessed with bottles. Do you think that may be the reason the application is having trouble mounting a fake ISO? Or is there a separate setting that's suppose to help with that in the preferences?

If you just name one of your drives to `FF7DISC1` ( or 2 or 3 at the end ) and you use `Play with Minimal Validation` you'll be able to run the game skipping the disk mount logic.

**Comment by PlUmPaSsChIcKeN** [comment]
> Seem to be running into the issue with the 7th heaven mounting the disco or even mounting the virtual disco when i swapped options as per mentioned earlier in the thread. The Bottle recipe calls for install steam then FF7 then 7th heaven. Given that all usually takes place within a fake C drive, does that mean that I would need to have FF7 be placed in the C drive created in bottles?
> 
> Given steam deck has it's built in steam application, it's made with it's own fake drive which can be accessed with bottles. Do you think that may be the reason the application is having trouble mounting a fake ISO? Or is there a separate setting that's suppose to help with that in the preferences?
> 
> ![ksnip_20220527-193154](https://user-images.githubusercontent.com/101153429/170800232-dbbcf58b-41a9-465c-b667-346b63f2abd5.png)

Hey, also a Steam Deck user that is interested in using 7th Heaven. Did you get yours to work with mods? If so, how did you get it to work? Any suggestions as I'm interested in trying this myself

**Comment by lumnie** [comment]
I did managed to get this to work mods included. It wasn't the easiest process for me as I'm pretty  inept when it comes to linux but basically it was a process of Creating a Bottle with the recipe provided above > Renaming one of the drives in the bottles program to FF7DISC1 and then running the game. Keyboard and mouse highly recommended for setting up directories for 7th heaven. I had ended up moving my installed FF7 out of it's default location to one much more manageable by 7th Heaven's recognition. Same with mods.

It plays okay. There's some settings I've yet to tweak as my game tends to run slow given my mod load. I think for the most part, it's a matter of having to toggle DXVK off and on when messing with 7th Heaven so that it functions properly and the fact I'm not able to add it to steam to work that way from game mode.

**Comment by PlUmPaSsChIcKeN** [comment]
That's awesome to hear. Did you install bottles via the terminal command posted above or was there any other steps involved?

**Comment by lumnie** [comment]
No terminal. I used the long one posted above and created a config file and then just imported the settings in Bottles when making a Custom Bottle based off that long string you can copy from sjenkins above. I tried the terminal but was getting confused as it wasn't properly launching the program database that bottle populates for me.

It's been a bit since I tried messing with it again so I might try again with figuring out how to get this to work in gaming mode.

**Comment by julianxhokaxhiu** [comment]
> That's really unfortunate about not being to launch in game mode. Wonder if there is a solution for this

What do you mean exactly?

**Comment by lumnie** [comment]
> > That's really unfortunate about not being to launch in game mode. Wonder if there is a solution for this
> 
> What do you mean exactly?

Steam Deck operates in gaming mode for their basic game launching ( Basically Steam Big Picture Mode ) but has a desktop mode for the traditional desktop experience where you can do all the fun backend stuff. Usually to get something to launch in gaming mode, you'd just add the program as a non-steam game in desktop mode and then you'd be able to launch the program from your library when you swap back to gaming mode.

Been trying to add bottles as a program shortcut and launch it that way to get to 7th Heaven and by proxy my modded FF7 but it doesn't seem to be a viable option that way. Haven't found a fix so just shelved it for now on my end.

**Comment by PlUmPaSsChIcKeN** [comment]
> > > That's really unfortunate about not being to launch in game mode. Wonder if there is a solution for this
> > 
> > 
> > What do you mean exactly?
> 
> Steam Deck operates in gaming mode for their basic game launching ( Basically Steam Big Picture Mode ) but has a desktop mode for the traditional desktop experience where you can do all the fun backend stuff. Usually to get something to launch in gaming mode, you'd just add the program as a non-steam game in desktop mode and then you'd be able to launch the program from your library when you swap back to gaming mode.
> 
> Been trying to add bottles as a program shortcut and launch it that way to get to 7th Heaven and by proxy my modded FF7 but it doesn't seem to be a viable option that way. Haven't found a fix so just shelved it for now on my end.

Did you install steam inside a custom bottle or were you able to use the pre-installed steam and then install FFVII from there?

**Comment by lumnie** [comment]
Correct. I just used the pre-installed steam to download FFVII. The guide only seemed like installing steam first was necessary so that you could get access to FF7 and with the deck, I felt like I could cut that part out. Also yes, I had the flatpack version of bottles installed.

**Comment by sjenkins7** [comment]
Apologies for the radio silence for a while - I've been super busy so unable to really catch up on things here.

It's increasingly looking like i'm back to step 1 again with this - Having pulled down the latest bottles and 7th heaven installer, its fairly broken again.

I'll start looking at getting things patched up and working on the installer branch I have, but it may take some time.

As an aside - it may pay for us to look into alternatives, such as providing our own setup via Flatpak. the FF14 community have done something very similar there, as they can now better control the libraries / runtimes. That said - not sure how that would work for us here, given we also need to source FF7 as well.

**Comment by julianxhokaxhiu** [comment]
Just coming by to ask what is the current situation around this: I've seen multiple users reporting on Discord playing FF7 w/7th on a Steam Deck. Is there still anything required on our own end or can we close this issue? LMK, thanks

**Comment by sjenkins7** [comment]
Work has been pretty busy so i've been unable to finish updating things. I've not been up-to-date with Discord, but that's some good news. I'll see what i can do about finishing updating stuff over this weekend.

I'd suggest not closing it until the launcher is public - it would also make the installation much much simpler for those involved

**Comment by sjenkins7** [comment]
Still no joy - for whatever reason the only runner that I could get working previously is no longer available. Every other one either causes 7th Heaven to crash on launch, or causes the "code 5" issue, of which the workaround doesn't work at all in wine.

Because `vaniglia` is unavailable, unsure how Steam Deck users are currently able to use this at all.

Been trying to avoid making a fully native launcher to help lock down all these moving pieces, because it would take a fair amount of time. It's increasingly looking like I might not have alot of choice, unless we can deal with the underlying hooking mechanism ( which i suspect would just break everyone's existing mods.... ). At very least, a native launcher may make it easier for the Steam Deck users to run in Big Picture / gaming mode.

Don't suppose anyone on the team has time to chat about this in Discord in some detail?

**Comment by PlUmPaSsChIcKeN** [comment]
Haven't checked the discord in a while but people have gotten 7th Heaven to work?

I was messing around with it a while ago and managed to get 7th Heaven installed using steam but was having errors whenever I added the game directory to it and tried to run it. I'm wondering if those users are running windows instead. If not, were they able to get it running in game mode? 

One of the major hassles with it was the on screen keyboard kept getting stuck and would glitch out so very difficult to maneuver around using only the steam deck. Definitely need to plug it into an external dock

**Comment by Hannan100** [comment]
Anyone had any luck with this recently? I tried following the earlier Bottle approach today without much success (got 7th installed, but it failed to launch even the vanilla game).

**Comment by SeongGino** [comment]
Still the same behavior, and will probably stay this way until it's directly addressed.

**Comment by cyberconan** [comment]
Hi everyone!
Thanks sjenkins7 for the advice with vaniglia wine version. At the moment on Archlinux I can play ff7 with the new 7th heaven releases after downloading the wine fork from [here](https://github.com/bottlesdevs/wine/releases/download/6.23/vaniglia-6.23-x86_64.tar.gz) and using a little script:

```
#!/bin/sh
ff7_path="<my ff7 game path>"
wine_bin="$ff7_path/vaniglia-6.23-x86_64/bin/wine"

export LC_ALL="C"
export WINEESYNC=1
export WINEPREFIX="$ff7_path/wine-ff7"

"$wine_bin" "7th Heaven.exe" /PROFILE:Default /LAUNCH /QUIT
```

Thanks also to The Tsunamods team for keeping the project alive.

**Comment by SeongGino** [comment]
Unfortunately, am unable to reproduce any working functionality with the linked Vaniglia fork:

```
Inconsistency detected by ld.so: dl-call-libc-early-init.c: 37: _dl_call_libc_early_init: Assertion `sym != NULL' failed!
Monitored process exited.
Initial process has exited (return code: 32512)
```
Using latest Arch glibc.
The newest Vaniglia release doesn't even have a `wine` executable so Lutris doesn't see it.

**Comment by cyberconan** [comment]
> Unfortunately, am unable to reproduce any working functionality with the linked Vaniglia fork:
> 
> ```
> Inconsistency detected by ld.so: dl-call-libc-early-init.c: 37: _dl_call_libc_early_init: Assertion `sym != NULL' failed!
> Monitored process exited.
> Initial process has exited (return code: 32512)
> ```
> 
> Using latest Arch glibc. The newest Vaniglia release doesn't even have a `wine` executable so Lutris doesn't see it.

To call the wine fork from Lutris you would have to clean up the old library calls in the folder. To fix this I use the script I posted that calls the compiled Vaniglia binary from bash using the archlinux system libraries instead of the ones it carries. Try to use the script and you will see how the 7th heaven.exe starts.

**Comment by SeongGino** [comment]
> To call the wine fork from Lutris you would have to clean up the old library calls in the folder. To avoid this I use the script I posted that calls the compiled Vaniglia binary from bash using the archlinux system libraries instead of the ones it carries. Try to use the script and you will see how the 7th heaven.exe starts.

Well from what I can tell, all your script does is set the `WINEPREFIX`, esync, whatever `LC_ALL` is, and runs the program using a designated wine version; all things Lutris does. I ended up setting the `Disable Lutris Runtime` setting for my 7th Heaven launcher... er, launcher--and Vaniglia works just fine.

Indeed, 7h hooks seems to work just fine with this specific Wine fork.

I experimented a bit, and the Vaniglia fork is the only one that seems to function with the hooks - allegedly it's "just" Wine Staging patches, but the corresponding Wine-Staging release throws a similar error to every other Wine release. But FWIW, Vaniglia 7.5 (i.e. the release *before* Vaniglia 7.4) also works. But the following release doesn't, and it seems every subsequent Vaniglia release strips out WoW64 support; hence the lack of a normal `wine` executable.

I'd also like to point out that using it with FSYNC enabled seems to not have any harmful effects. And now I'm wondering what the point was of using the `LC_ALL=C` flag...

P.S. Using this Wine version causes 7h to crash whenever clicking on anything that renders a dropdown menu, as well as simply hovering over the status bar/"click here to display logs" area, which is very easy to do by sheer accident.

**Comment by Hannan100** [comment]
For what it's worth, I was eventually able to get everything working on Steam Deck once I made sure I was using the `vaniglia` runner. The other stumbling block was setting the drive label with wineconfig not working, but I eventually got around that by placing a `.wimdows-label` file in the drive directory.

Everything now works: 7H, vanilla FF7 and modded FF7. The only thing I can't get to work is the Bottles shortcut in Steam to access it in the Deck UI rather than Desktop mode, but that's not the end of the world.

**Comment by aronjanosch** [comment]
That's excellent progress. Can you give a small step-by-step guide on how to get it working?

**Comment by selektionsrest** [comment]
Got Bottles setup and 7th Heaven launches.... but when i start FF7 it crashes, even without mods....

**Comment by persunde** [comment]
@lumnie 
Do you or anyone else have any tips for how I can get bast the mounted ISO requirement.
I am just not able to get past that requirement...

I have been unable to add a fake drive called FF7DISC1 that Proton will see, so running with Minimal Validation also does not work :(

**Comment by persunde** [comment]
Is there a reason why 7th Heaven requires the ISO to be mounted?
I can easily play FF7 on Steam using proton, but it does not work when running 7th heaven on Proton.

**Comment by SeongGino** [comment]
Because the original legal copy has a CD check that requires this workaround. The Steam release is a whole different game technically.

The alternative would be piracy, which I'm just going to guess shouldn't be discussed on a public forum like this. But even then, I doubt the alternative method would work.

The marking a disc as `FF7DISC1` method works fine in normal Wine.

**Comment by persunde** [comment]
Is it possible for this mod to run using the steam release .exe file for those who want to use that?
I believe that should solve the problem and make it very easy to run 7th heaven on linux.

It was super easy to run FF9 with the moguri mod using Stream Proton. But because 7th heaven runs using the non-steam .exe that requires disk verification it is very complicated to get it running.

@Hannan100 
Can you please describe in more detail how you were able to set the label on the ISO disk.
I tried to make a dummy ISO like this:
`mkisofs -o /tmp/FF7DISC1.iso -V FF7DISC1 /tmp/empty-file`

I managed to mount it using protontricks, but 7th heaven just wont accept the CD. I believe the drive label is not set, but I am not sure.

**Comment by sjenkins7** [comment]
> I managed to mount it using protontricks, but 7th heaven just wont accept the CD. I believe the drive label is not set, but I am not sure.

Make sure you have .NET 4.8 installed. There appears to be an issue with mono, where it doesn't return the correct disc label.

**Comment by julianxhokaxhiu** [comment]
Hey community, I'd like to let you know about the imminent release of what is going to be the 7th Heaven 3.0 release. If you want to know more, you'll find the announcement on the qhimm discord server, but in a nutshell is a .NET 7 version with a completely rewritten injection engine that will work without admin permissions or code 5 issues. Feel free to try the latest canaries and let me know how it goes!

**Comment by SeongGino** [comment]
For what it's worth, the .NET 7 canary build works *better* with Linux, so long as the launcher is loaded with a `DOTNET_ROOT=c:\Program Files\dotnet\\` environment variable for it to not confuse the native Linux version with the Windows one it installs. It still needs either DXVK disabled or a dxvk.conf file in the 7H directory with
```
[7th Heaven.exe]
d3d9.shaderModel = 1
```
to render the window properly, but otherwise it works from 7H's installer on a fresh prefix.

Unfortunately though, trying to run FF7 has been a challenge; on a new prefix using only the .NET and MSVC redists the installer includes, trying to run the game (modded or vanilla) causes FF7 to ignore FFNx and loads its software renderer because it's not respecting any game driver settings and even messed with my desktop resolution, then immediately crashes (with 7H still open). At least running modded isn't technically generating any errors now?

I'm also noticing it keeps trying to re-register registry entries that are already present?

This is with Wine-GE.

**Comment by sjenkins7** [comment]
After a little bit of playing around last night & this morning  - I was able to get the game launched with the new Canary, but no mods appear to take effect:

![Screenshot_20221212_083104](https://user-images.githubusercontent.com/20030065/206999789-15c8dbf9-6d6d-42d5-8175-187b4c305c06.png)

I could only get this far with Proton 7.0 currently - using WINE 7.0.1 from flatpak currently just segfaults on launching 7th Heaven.

Closing the game currently causes it to freeze & the 7th Heaven Canary doesn't respond to closing the window or ALT + F4.

I have seen the same issue as @SeongGino while building this from clean - I'm wondering if there is something different being done in the Canary on a clean install that isn't taking effect if you upgrade from Stable?

> It still needs either DXVK disabled or a dxvk.conf file in the 7H directory with

Thanks for this - trying to disable DXVK when Flatpak bundles it by default has proven difficult. Should go a long way to resolve the glitchy / black-screen rendering i was having.

**Comment by sjenkins7** [comment]
In addition - Had to install mods & FFNx manually. For some reason 7th Heaven is unable to check for updates or read the catalog

**Comment by julianxhokaxhiu** [comment]
> I'm wondering if there is something different being done in the Canary on a clean install that isn't taking effect if you upgrade from Stable?

Yes, upgrading 7th using the internal update would not install the .NET 7 Desktop Runtime required to run it, as well as the VC++ Runtime for the native injection to work.

> I was able to get the game launched with the new Canary, but no mods appear to take effect:
Not sure how to check that under Wine but you need to make sure hooking Win32 APIs is allowed as we use Detours in order to do so. If there's a way to make it verbose please check the logs if any error or warning shows up there.

> For some reason 7th Heaven is unable to check for updates or read the catalog
If you used the setup to install it and you did on a usually Administrator based path ( Program Files ) I pushed a patch that will fix this permission issues ( at least on Windows ). If you're using the ZIP and installing it on your own path then I have no idea, again check the logs in either 7th or Wine to see if anything shows up. I'd start from the 7th `applog.txt` and moving around it.

**Comment by sjenkins7** [comment]
> Yes, upgrading 7th using the internal update would not install the .NET 7 Desktop Runtime required to run it, as well as the VC++ Runtime for the native injection to work.

Just to be clear - I am still running the launcher itself over the top of a working ( vanilla only ) Stable install. It automatically installed the runtimes as needed.

> Not sure how to check that under Wine but you need to make sure hooking Win32 APIs is allowed as we use Detours in order to do so. If there's a way to make it verbose please check the logs if any error or warning shows up there.

I'll re-run this and try to get some decent logs. Didn't have time to capture everything this morning as it was a quick test before work.

> If you used the setup to install it and you did on a usually Administrator based path ( Program Files ) I pushed a patch that will fix this permission issues ( at least on Windows )

This is the `2.9.9.44` install, which has this included? Or are you referring to an additional change since?

**Comment by julianxhokaxhiu** [comment]
2.9.9.44 has the change I mentioned so you should be good, but only the setup ( the .exe download on the asset list basically )

**Comment by julianxhokaxhiu** [comment]
Tbh looks like a network stack issue than a 7th issue on its own. Maybe check if there's anything related to this issue on Wine forums. I'll also try to dump some more details on the logs so the next time we can see where and why it fails.

**Comment by sjenkins7** [comment]
Yeah i suspect a chunk of this will be down to the Switch to .NET 7 and how that is handled in Wine. I've been digging around since but so far, we appear to be one of the only ones using such a new version of .NET at this time.

Watching the debug output for Wine has it's usual `fixme` logs, but didn't see anything related. I'll look into how winetricks handles older .NET installs, as I recall it being somewhat of a pain.

**Comment by julianxhokaxhiu** [comment]
Yeah .NET 7 has been recently released so I can imagine the struggle to get it running on Wine but I'm confident the community will find their way out. I honestly don't have time to run in parallel on this as I'm really overbooked with thingas to follow and next things to work on around FFNx and 7th, but I'll be happy to provide any patch required to help on this mission. Patience and willingness is what will make this port eventually succeed. Don't give up!

**Comment by sjenkins7** [comment]
Thanks for that - re-ran with the .49 canary and seeing a number of the same underlying exception:

`System.Runtime.InteropServices.SEHException: External component has thrown an exception.`

Added the full log, but there isn't alot else there: 
[applog.txt](https://github.com/tsunamods-codes/7th-Heaven/files/10212912/applog.txt)

I don't see an error code to go with this, so unsure how to help narrow this down

**Comment by julianxhokaxhiu** [comment]
Looks like some kind of unmanaged exception from .NET perspective: https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.sehexception?view=net-7.0#remarks

I have no idea which component is used to download but I can try to check later what is going on and If I can understand something around it. Although as it works fine under Windows I might think it's something related to Wine and its underlying layer :(

**Comment by sjenkins7** [comment]
Thanks for the extra info in the edit - Will be very useful. The HTTP Listener part seems interesting, as the flag implies there may be other parts of .NET that could potentially be forced into a Managed implementation, which is likely to be better in terms of compatibility.

I have access to a Windows VM, so I'll download the source and go playing around - see what comes out of it

**Comment by julianxhokaxhiu** [comment]
Actually, one of the warnings I get from .NET 7 now is that the core component we use to do downloads is deprecated. I'll try to see if I can easily migrate to what .NET suggests, this might fix also your issue perhaps.

**Comment by julianxhokaxhiu** [comment]
I was able to migrate one of the basic download classes, so most of the catalog should work fine. Google Drive links will still use the old download code until I figure out how to migrate that. Feel free to test the latest canary.

**Comment by SeongGino** [comment]
Catalog does seem to work at least in latest canary.

Interestingly, I find that the catalog downloader module crashes with current Wine-GE builds (External component as thrown an exception), but using a build of current Wine Staging (which is my system's Wine) allows it to work.

**Comment by julianxhokaxhiu** [comment]
Were you able to download some mods? What about launching the game? Were you able to load them?

**Comment by sjenkins7** [comment]
Quick re-run of `v2.9.9.58` was the same as before for me - both FFNx updates & mod catalogs failed with the same exception.

**Comment by julianxhokaxhiu** [comment]
@sjenkins7 I suppose it has nothing to do with .NET then as @SeongGino was able to run it...double check if you're both on the same Wine codebase or something...please also remember that if you use different distributions you might have different patch sets applied on top. Safer option would be to test this on Arch linux.

**Comment by sjenkins7** [comment]
Agreed @julianxhokaxhiu  - Staging seems to be the way to go, but I don't have this as an option for Flatpak currently. I'll re-try it later on my main PC, rather than on the Steam Deck

**Comment by julianxhokaxhiu** [comment]
Yes please, let's focus on a mainline Linux first so it will be easier to pinpoint eventual Steam Deck issues.

**Comment by sjenkins7** [comment]
OK wine Staging 7.22 - Able to get the catalog now.

Attempting to launch the game out-of-the-box fails, as the registry isn't quite setup right:

```
[Software\\Wow6432Node\\Square Soft, Inc.\\Final Fantasy VII] 1671188680
#time=1d9113e319f62e8
"AppPath"="C:\\Games\\FINAL FANTASY VII\" /t REG_SZ /f /reg:64"
"DataDrive"="X:\" /t REG_SZ /f /reg:64"
"DataPath"="C:\\Games\\FINAL FANTASY VII\\data\" /t REG_SZ /f /reg:64"
"DiskNo"=dword:00000000
"FullInstall"=dword:00000001
"MoviePath"="C:\\Games\\FINAL FANTASY VII\\data\\movies\" /t REG_SZ /f /reg:64"
```

Explains why i'm being greeted with multiple command line prompts asking to override the key on each launch attempt

**Comment by sjenkins7** [comment]
Just to add - If i manually correct those entries, I get to the same point as @SeongGino previously - Game fullscreens & immediately crashes.

Have to manually reset the display resolution after

**Comment by SeongGino** [comment]
@sjenkins7 and that's why I've been running in virtual desktop mode, for what it's worth.

> Were you able to download some mods? What about launching the game? Were you able to load them?

I actually was able to download a mod (Finishing Touches, just as an example) with Wine Staging (Wine version seems to be a root cause in this, since Wine-GE fails at even downloading previews even with a pre-fetched catalog). The download bar itself doesn't immediately indicate progress, it stayed at 0% and estimated time at pending, but then went right to 100% after a few moments. After which it installed fine and showed up in the mod list, and I can configure it with no corruption or anything.

Er, still same launch issue though. Registry override popups (like Jenkins earlier), then game still only loads the old renderer and crashes outright, whether vanilla or not.
FWIW I can go back to my 7H 2.4.1.0 prefix [detailed earlier here](https://github.com/tsunamods-codes/7th-Heaven/issues/19#issuecomment-1223502966) running that older stable version, using the same instance of FF7 and it'll load FFNx and mods fine. But the new canary version in its own prefix, no dice when actually loading the game thus far.

**Comment by julianxhokaxhiu** [comment]
May I ask if you can please post the Wine config you're using?

**Comment by SeongGino** [comment]
Running Arch and doing all this through Lutris, fwiw:

Wine-staging from Arch multilib repos
Using Lutris's defaults with the exception of Windowed (Virtual Desktop) enabled, and an environment variable `DOTNET_ROOT=c:\Program Files\dotnet\\`. All running on a clean prefix using nothing but the dependencies provided from 7H Canary's .exe installer. Then I set a custom *D:\* pointing to an install of FF7's directory using the label `FF7DISC1`.

**Comment by sjenkins7** [comment]
Spent a bit of time diffing my previously working (stable but vanilla game only ) prefix & one created using the latest canary - seems that there was 2 values missing from `system.reg`:

```
[Software\\Wow6432Node\\Square Soft, Inc.\\Final Fantasy VII\\1.00\\Graphics]
"DD_GUID"=hex:00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00

[Software\\Wow6432Node\\Square Soft, Inc.\\Final Fantasy VII\\1.00\\Sound] 
"Sound_GUID"=hex:00,00,00,00,00,00,00,00,00,00,00,00,00,00,00,00
```

Both appear to be needed to get Graphics & sound to work at all.

I get back to roughly where I was before - Game freezes if you quit / ALT+F4. I can launch with mods, but none appear to be taking effect ( I picked what i felt to be obvious ones to be sure ).

![image](https://user-images.githubusercontent.com/20030065/209549538-e7e8de1f-0c5d-4859-ae6d-69776770190c.png)

**Comment by sjenkins7** [comment]
Possibly related - Missing logs cause 7th Heaven to crash, if FF7 exits uncleanly: 
![image](https://user-images.githubusercontent.com/20030065/209550205-48358dd5-0b8e-4691-996b-7498bdd9bc1a.png)

On the face of it, seems like the wrapper is just silently failing?

**Comment by SeongGino** [comment]
Turns out in my case, it was both the missing keys that @sjenkins7 referenced, and that my AppPath/DataPath/MoviePaths were somehow malformed (which is why 7H kept prompting me to change the registry, because it turns out my paths somehow turned into
![2022_12-26-1127411](https://user-images.githubusercontent.com/7321839/209568853-0656d373-22e0-4f06-8bb8-914c38bdc54d.png)
didn't matter if it was a virtual B:\ or Z:\ ; it always does this)
After manually fixing those registry keys, *now* FF7 boots up fine with FFNx.

But yeah, mods still don't load at all. My guess is it would be the wrapper failing?

**Comment by sjenkins7** [comment]
@SeongGino I believe the registry mangling was resolved in the latest / `.71` canary. I did mine from scratch just to be sure

**Comment by SeongGino** [comment]
@sjenkins7 I was using `.61`; I guess I just missed on those fixes from the work I'd been doing the past couple of weeks. Good to know!

**Comment by sjenkins7** [comment]
Was able to get this working on the latest canary (`.86`). Make sure you configure `dinput` correctly

```
# install dinput native DLL
winetricks dinput
```

to make sure - go into `winecfg` and set `dinput` to be `native`

![image](https://user-images.githubusercontent.com/20030065/212287327-42df5cb1-3640-4107-bca3-4f00f2eba520.png)

**Comment by sjenkins7** [comment]
Also able to confirm that these additional bits would be enough to get this working on the Steam Deck.

In terms of packaging this up nicely ( Flatpak ), we are currently stuck waiting for a stable release of wine ( `7.0.1` is the current & doesn't work at all )

There are 2 avenues right now:

* There is talk about adding development versions to Flatpak ( which would serve what we need ) - https://github.com/flathub/org.winehq.Wine/issues/20
* We wait for 8.0.0 and see if that works ( they are on RC4 right now, so I can't imagine it will take long ).

**Comment by julianxhokaxhiu** [comment]
I updated the priority as a reminder for a follow up. Great job there, looking forward to an easy installer.

**Comment by SeongGino** [comment]
Please advise on how to resolve this popup when attempting to start modded FF7 on latest canary & wine-staging.
![2023_01-14 Error-123958](https://user-images.githubusercontent.com/7321839/212487656-71e45458-3a88-47c5-a796-288311f13e52.png)
Aside from `dinput` through Winetricks, prefix is only using what 7H canary installer provides.

**Comment by sjenkins7** [comment]
> Seems like 7H is unable to find the 32-bit half of .NET 7.0.1, hence the error in launching w/ mods despite being able to run the launcher.

Hopefully these kind of things will be less of an issue going forward, as wine shouldn't be able to see the native install in the proposed Flatpak

**Comment by ChthonVII** [comment]
May I ask for a recap of this very long thread? It sounds like there have been fully working configurations at a couple points, but it's not super clear what's actually needed at this point.

It sounds like the following should be a working configuration. Please correct what's wrong and missing:
* Wine staging 8.0rc4 (official build)
* 64-bit wine prefix?
* 7th Heaven canary .86
* native dinput via winetricks
* native dotnet4 and 48 via winetricks?
* Unclear which things mentioned in the March post are still needed

**Comment by sjenkins7** [comment]
There are 2 trains ongoing, which is likely where the confusion is coming from. For simplicity sake, I'll focus just on the Canary, as this will eventually become the new stable.

* Wine-staging 7.22
* 64-bit prefix
* 7th Heaven Canary `(.86)`
* the following installed via winetricks:
```
    winetricks -q d3dx9
    winetricks -q msls31
    winetricks -q riched20
    winetricks -q corefonts
    winetricks -q d3dcompiler_43
    winetricks -q d3dcompiler_47
    winetricks -q dinput
```
* .NET Desktop Runtime 7.0.1+ ( both x64 & x86 versions)
* Virtual CD configured ( make sure it is explicitly marked as a CD-ROM in `winecfg` ):
```
    mkdir -p "${WINEPREFIX}/drive_c/FF7DISC1"
    echo "FF7DISC1" > "${WINEPREFIX}/drive_c/FF7DISC1/.windows-label"
    echo "44000000" > "${WINEPREFIX}/drive_c/FF7DISC1/.windows-serial"
    touch "${WINEPREFIX}/drive_c/FF7DISC1/FF7DISC1.TXT"
    ln -sfv "${WINEPREFIX}/drive_c/FF7DISC1" "${WINEPREFIX}/dosdevices/x:"
```

If you have .NET 7.0 installed natively, it may interfere ( see https://github.com/tsunamods-codes/7th-Heaven/issues/19#issuecomment-1382869290 )

Please note that generally speaking - closing FF7 will cause the app to hang.
    
Any further talk in here is likely to be about packaging & potentially usable wine versions ( because staging is not available as a Flatpak, which is currently stalling progress )

**Comment by dotaxis** [comment]
Running 7th Heaven Stable with Bottles:
https://www.youtube.com/watch?v=ZtZNSPygxyY

Running 7th Heaven Canary with GE-Proton:
https://www.youtube.com/watch?v=Wfh-pE_TvGw

**Comment by SeongGino** [comment]
>     * the following installed via winetricks:
> 
> 
> ```
>     winetricks -q d3dx9
>     winetricks -q msls31
>     winetricks -q riched20
>     winetricks -q corefonts
>     winetricks -q d3dcompiler_43
>     winetricks -q d3dcompiler_47
>     winetricks -q dinput
> ```

FWIW, I haven't needed to do any of the noted above Winetricks in the clean prefix I made, apart from `dinput` - the prereqs provided from/installed by the Canary installer (and having no native Linux .NET 7 install) makes it work out of the box with either Wine-GE or Wine-Staging (with just the former crashing the catalog downloader module for whatever reason, at least until it's updated?). Though this is on Lutris w/ DXVK & VKD3D installed in the prefix.

**Comment by ChthonVII** [comment]
@sjenkins7:

I followed you instructions with minor deviations, but I'm afraid 7H crashes with a stack overflow when I try to execute it:
`0108:err:virtual:virtual_setup_exception stack overflow 1808 bytes addr 0x1d36898 stack 0x208f0 (0x20000-0x21000-0x1a0000)`

Deviations:
- Wine is latest staging, 8.0rc4
- 7H is the new canary that dropped a few hours ago, v2.9.9.89 (I'd love to try .86, since this is likely the culprit, but it seems to have disappeared from the releases page.)
- .NET Desktop Runtime is 7.0.2

If anyone could toss up a .86 download somewhere, I'd appreciate it.

Also, do I absolutely _need_ 7H? Is it possible to manually unpack the iro files into FFNx's mod directories and achieve the same result? Or does 7H have some functionality above and beyond what FFNx does? (On further thought, it would probably work better to unpack them elsewhere and then symlink stuff into the FFNx directories. That way I could simulate a "load order" change without having to re-unpack stuff that had been overwritten.)

**Comment by julianxhokaxhiu** [comment]
> Also, do I absolutely need 7H? Is it possible to manually unpack the iro files into FFNx's mod directories and achieve the same result? Or does 7H have some functionality above and beyond what FFNx does? 

7th has an internal way to ensure you load files based on how many mods you load, and there are some dynamic behavior done that mods do use to provide special functionalities. So dropping files manually unfortunately won't make the mods work as you might expect.

Related to the 7th version, nothing critical changed between .86 and .89. I suspect the major difference here is Wine. Consider using the version suggested by @sjenkins7 and feel free to join the Discord channels for better support.

**Comment by sjenkins7** [comment]
> I followed you instructions with minor deviations, but I'm afraid 7H crashes with a stack overflow when I try to execute it:
0108:err:virtual:virtual_setup_exception stack overflow 1808 bytes addr 0x1d36898 stack 0x208f0 (0x20000-0x21000-0x1a0000)

I've just run `.89` locally to make sure & it launches fine using:

```
$ wine --version
wine-7.22 (Staging)
```

I would suggest making sure you use this for now. I'm not aware of any distribution that has decent staging packages beyond this yet.

If I get time, it might prove useful for me to understand the staging patching process, so I can narrow down which patch(es) we require, in order to know when we can switch to a normal / stable build. On the face of it, that looks like a pretty big task though, so that will take some time.

**Comment by SeongGino** [comment]
Also checked to make sure, using versions that report as `wine-7.22 (Staging)` and `wine-8.0rc4.r0.g89374093 ( TkG Staging Esync Fsync )` (in Arch these are `wine-staging` from multilib and `wine-tkg-staging-fsync-git` from AUR, respectively), 7H and modded FF7 are definitely working fine as of current Canary.

**Comment by ChthonVII** [comment]
Still getting nowhere with 7.22

```
export WINEPREFIX=~/.wine-ff7h
export WINEVERPATH="/home/<username>/Programs/wineforks/vanilla722/dist"
export WINELOADER="/home/<username>/Programs/wineforks/vanilla722/dist/bin/wine"
export WINESERVER="/home/<username>/Programs/wineforks/vanilla722/dist/bin/wineserver"
export WINEDLLPATH="/home/<username>/Programs/wineforks/vanilla722/dist/lib64/wine":"/home/<username>/Programs/wineforks/vanilla722/dist/lib/wine":"/home/<username>/Programs/wineforks/vanilla722/dist/lib64/wine/x86_64-windows":"/home/<username>/Programs/wineforks/vanilla722/dist/lib/wine/i386-windows"
export LD_LIBRARY_PATH="/home/<username>/Programs/wineforks/vanilla722/dist/lib64":"/home/<username>/Programs/wineforks/vanilla722/dist/lib":"/home/<username>/Programs/wineforks/vanilla722/dist/lib64/wine/x86_64-unix":"/home/<username>/Programs/wineforks/vanilla722/dist/lib/wine/i386-unix"
export PATH="/home/<username>/Programs/wineforks/vanilla722/dist/bin:$PATH"
export WINE="/home/<username>/Programs/wineforks/vanilla722/dist/bin/wine"
wine --version
<output: wine-7.22>

winecfg
<prefix is created>
cd .wine-ff7h/drive_c/
mkdir Temp
cd Temp
cp ~/Installs/WINE/FF7/7thHeaven/v2fork/windowsdesktop-runtime-7.0.2-win-x* .
cp ~/Installs/WINE/FF7/7thHeaven/v2fork/7thHeaven-v2.9.9.89_Release.exe .
wine windowsdesktop-runtime-7.0.2-win-x86.exe
<click through installer>
wine windowsdesktop-runtime-7.0.2-win-x64.exe 
<click through installer>
wine 7thHeaven-v2.9.9.89_Release.exe
<click through installer>
<leave "run 7th Heaven" checkbox checked>
<crash: 01cc:err:virtual:virtual_setup_exception stack overflow 1808 bytes addr 0x1d46898 stack 0x208f0 (0x20000-0x21000-0x1a0000)>
```

**Comment by sjenkins7** [comment]
You are using a non-staging release, which is likely why it doesn't work.

Need to make sure that `wine --version` specifically reports `(Staging)`. 

If you do not have a distribution that packages this for you, then you will have to build it yourself sadly - As far as I know, the WineHQ website doesn't have premade packages for this

**Comment by ChthonVII** [comment]
> You are using a non-staging release, which is likely why it doesn't work.
Yep. That was it. I grabbed the devel binaries by mistake. Always thought devel had the staging patches, though... Guess not.

> If you do not have a distribution that packages this for you, then you will have to build it yourself sadly - As far as I know, the WineHQ website doesn't have premade packages for this
They've got binaries for at least some distros. [Here](https://dl.winehq.org/wine-builds/debian/dists/bullseye/main/) are the Debian files. (Just use dpkg-deb to extract the two i386 packages to one directory, and the two x86-64 packages to another directory, then cut/paste the x86-64 directory over the top of the i386 one, overwriting conflicts.)

I'm seeing 3 bugs right now, none of them fatal:
1. FF7 always hangs on exit.
2. Sometimes 7H renders its modal dialogs with black text/buttons on a black background, which is unusable. (This is using the light theme, so it's the background that's wrong.) Closing and restarting the program restores the correct behavior.
3. The DX11 renderer does not work out of the box. The opening movies have incorrect colors (seems like the wrong pixel format) and the first menu screen is a jumbled mess. I'm guessing this is due to the lack of DVXK. OpenGL and Vulkan both appear to work fine though (at least for the first 10 minutes or so that I've tested). Is there any reason to prefer DX11?

**Comment by julianxhokaxhiu** [comment]
> 1. FF7 always hangs on exit.

Update FFNx to the latest canary.

**Comment by SeongGino** [comment]
> 3. The DX11 renderer does not work out of the box. The opening movies have incorrect colors (seems like the wrong pixel format) and the first menu screen is a jumbled mess. I'm guessing this is due to the lack of DVXK. OpenGL and Vulkan both appear to work fine though (at least for the first 10 minutes or so that I've tested). Is there any reason to prefer DX11?

@ChthonVII Assuming you're not using Lutris to set up the prefix, that would be correct. I'm using the Vulkan backend, but another suggested the DX12 backend with VKD3D if you have stability problems.

**Comment by julianxhokaxhiu** [comment]
I'll close this issue as the hardest part to make it work has now been done and it has been proved to be effective. Anything else regarding Linux packaging and so on will be continued on Discord. Thanks to everyone for the effort in achieving this huge milestone!

**Comment by bravely-beep** [comment]
The recent update to use .NET 8 appears to have broken Wine support. After installing and booting the 7th Heaven application for the first time, the General Settings window opens, but the window contents are entirely black, rendering it unusable.

**Comment by SeongGino** [comment]
I'd received a notif about someone saying the .NET8 update not working.

Just coming in to say: still works here.
![2024_03-01 114744-7th Heaven v3 4 0 0 - Mod Manager for Final Fantasy 7 ( Maintained by the Tsunamods Team )  Default  ](https://github.com/tsunamods-codes/7th-Heaven/assets/7321839/c4b92a67-43f1-4a5a-9432-c2f9822bb0b5)
![2024_03-01 115134](https://github.com/tsunamods-codes/7th-Heaven/assets/7321839/5dd12c09-efde-45dd-8800-de38dd5727f6)
